### ISSUE TITLE
systemd-networkd sets the same IPv4 address via DHCP on different machines

### ISSUE THREAD

**Issue by asheplyakov** [systemd/systemd#9609]
**systemd version the issue has been seen with**
> systemd 237
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > ALTLinux

**Expected behaviour you didn't see**
 > …

**Unexpected behaviour you saw**
 > Several machines with same hardware attached to Ethernet network obtain the same IP address via DHCP (MAC addresses of NICs are definitely different)

**Steps to reproduce the problem**

- Attach several machines with the same hardware to the Ethernet network with a DHCP server
- Configure machines to obtain IPv4 address via DHCP with systemd-networkd
- Reboot machines

Expected results: machines obtain unique IPv4 addresses
Actual result: machines obtained the very same IPv4 address

It's possible to reproduce the issue with (libvirt/kvm) virtual machines.
Each VM has a single virtio Ethernet adapter attached to the `default` libvirt network.

$ udevadm info -q all /sys/class/net/eth0
P: /devices/pci0000:00/0000:00:03.0/virtio0/net/eth0
E: DEVPATH=/devices/pci0000:00/0000:00:03.0/virtio0/net/eth0
E: ID_BUS=pci
E: ID_MODEL_FROM_DATABASE=Virtio network device
E: ID_MODEL_ID=0x1000
E: ID_NET_NAME_MAC=enx5254001634e9
E: ID_NET_NAME_PATH=enp0s3
E: ID_NET_NAME_SLOT=ens3
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Red Hat, Inc.
E: ID_VENDOR_ID=0x1af4
E: IFINDEX=2
E: INTERFACE=eth0
E: MATCHADDR=52:54:00:16:34:e9
E: MATCHIFTYPE=1
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/eth0
E: TAGS=:systemd:
E: USEC_INITIALIZED=1720199

$ cat /etc/systemd/network/eth0.network 
[Match]
Name=eth0

[Network]
DHCP=yes

[DHCP]
UseMTU=yes
UseDomains=yes
UseNTP=yes

**Analysis**

The problem seems to stem from systemd-networkd trying to make the (IPv4) address stable accross reboots (without having any extra state). 
However systemd-networkd does not use MAC address of the device being configured, instead it derives "unique" identifier from udev info, see 
https://github.com/systemd/systemd/blob/6dd9477dde37f3ecb7be6343543d3f30a9a40a10/src/libsystemd-network/dhcp-identifier.c#L73-L100

That data is indeed unique within a single machine, however it's not unique across the L2 network. In particular DUID generated by `dhcp_identifier_set_iaid` is the same for two machines with the same hardware. Thus attaching such machines to the same (IP) network results in duplicate client IDs and duplicate IP addresses.

**Checking the hypothesis**

 gdb --silent /lib/systemd/systemd-networkd
(gdb) break ../src/libsystemd-network/network-internal.c:52
Breakpoint 1 at 0x5555555a45d8: file ../src/libsystemd-network/network-internal.c, line 52.
(gdb) run

Breakpoint 2, net_get_name (device=device@entry=0x5555557eccb0)
    at ../src/libsystemd-network/network-internal.c:52
52                      if (name)
(gdb) print name
$1 = 0x0
(gdb) cont
Continuing.

Breakpoint 1, net_get_name (device=device@entry=0x5555557eccb0)
    at ../src/libsystemd-network/network-internal.c:52
52                      if (name)
(gdb) print name
$2 = 0x5555557f1830 "ens3"

(gdb) bt
#0  net_get_name (device=device@entry=0x5555557eccb0)
    at ../src/libsystemd-network/network-internal.c:52
\#1  0x00005555555afec4 in dhcp_identifier_set_iaid (ifindex=2, 
    mac=mac@entry=0x5555557f16e0 "RT", mac_len=6, _id=_id@entry=0x5555557f1703)
    at ../src/libsystemd-network/dhcp-identifier.c:114
\#2  0x0000555555599425 in sd_dhcp_client_set_iaid_duid (client=0x5555557f1600, 
    iaid=0, duid_type=<optimized out>, duid=0x0, duid_len=0)
    at ../src/libsystemd-network/sd-dhcp-client.c:382
#3  0x0000555555595716 in dhcp4_configure (link=link@entry=0x5555557ef7b0)
    at ../src/network/networkd-dhcp4.c:770
#4  0x000055555558b699 in link_configure (link=0x5555557ef7b0)
    at ../src/network/networkd-link.c:2652
#5  link_initialized_and_synced (rtnl=<optimized out>, m=<optimized out>, 
    userdata=0x5555557ef7b0) at ../src/network/networkd-link.c:2783
#6  0x00007ffff7abaac4 in process_reply (m=<optimized out>, 
    rtnl=0x5555557e4640) at ../src/libsystemd/sd-netlink/sd-netlink.c:367
#7  process_running (ret=0x0, rtnl=0x5555557e4640)
    at ../src/libsystemd/sd-netlink/sd-netlink.c:422
#8  sd_netlink_process (rtnl=0x5555557e4640, ret=ret@entry=0x0)
    at ../src/libsystemd/sd-netlink/sd-netlink.c:452
#9  0x00007ffff7abada3 in time_callback (s=<optimized out>, 
    usec=<optimized out>, userdata=<optimized out>)
    at ../src/libsystemd/sd-netlink/sd-netlink.c:759
#10 0x00007ffff7aac37a in source_dispatch (s=s@entry=0x5555557e60e0)
    at ../src/libsystemd/sd-event/sd-event.c:2311
#11 0x00007ffff7aadb2e in sd_event_dispatch (e=<optimized out>, 
    e@entry=0x5555557e3dd0) at ../src/libsystemd/sd-event/sd-event.c:2663
#12 0x00007ffff7aadd53 in sd_event_run (e=<optimized out>, 
    e@entry=0x5555557e3dd0, timeout=timeout@entry=18446744073709551615)
    at ../src/libsystemd/sd-event/sd-event.c:2723
#13 0x00007ffff7aadf67 in sd_event_loop (e=<optimized out>)
    at ../src/libsystemd/sd-event/sd-event.c:2744
#14 0x0000555555561334 in main (argc=<optimized out>, argv=<optimized out>)
    at ../src/network/networkd.c:158

That is, client ID is derived from "ens3" string on all machines.

**Comment by zdzichu** [comment]
Have you checked different values of `ClientIdentifier=` and `DUIDType=`?

**Comment by asheplyakov** [comment]
TL;DR version: systemd-networkd reuses the code which generates (DHCPv6) IAID to generate DHCPv4 client id. However 

- DHCPv4 client identifier must be unique to that client within the subnet.
- IAID must be unique within the IA bindings of **the given client**, in other words, there's nothing wrong with reusing IAIDs by different machines.

**Comment by asheplyakov** [comment]
@zdzichu setting ClientIdentifier=mac solves the problem, however it would be nice to get the valid IP address "out of the box"

**Comment by asheplyakov** [comment]
> TL;DR version: systemd-networkd reuses the code which generates (DHCPv6) IAID to generate DHCPv4 client id. 

Spoken too fast. In fact systemd-networkd uses both IAID and DUID to produce DHCPv4 client id, however for some reason the DUID is empty:

(gdb) frame 
#0  dhcp4_configure (link=0x5555557ef7b0)
    at ../src/network/networkd-dhcp4.c:671
(gdb) print link->network->duid.type
$7 = _DUID_TYPE_INVALID
(gdb) print link->manager->duid
$8 = {type = DUID_TYPE_EN, raw_data_len = 0 '\000', 
  raw_data = '\000' <repeats 127 times>}

**Comment by yuwata** [comment]
When `DUIDRawData=` is not set, `networkd` uses the value based on the machine ID.

**Comment by yuwata** [comment]
@asheplyakov So, I guess that your machines have the same machine ID. Is it true?

**Comment by asheplyakov** [comment]
@yuwata 
> I guess that your machines have the same machine ID

Apparently yes. `/etc/machine-id` is the same on all machines. However the image used to provision test VMs does not contain that file.

**Comment by asheplyakov** [comment]
So the problem is caused by a wrong configuration. However systemd-networkd could use a safer defaults, that is:

- Use `ClientIdentifier=mac` for DHCPv4
- Use DUID-LL for Ethernet and wireless devices for DHCPv6

Such a configuration is much more difficult to get wrong (unless one has got duplicate MACs within the same L2 network).

**Comment by ageric** [comment]
Having just run into this problem, I'd say that DHPv3 on IPv4 networks should definitely just use the MAC address to obtain it's IP address. Having the same MAC for two different IP's would cause similar issues anyway, and it's the first thing people with any experience will look for and should be the first thing people trying to describe the issue to google should find a solution for.

**Comment by b1tninja** [comment]
You mentioned vms, I believe maybe you need to run systemd-firstboot --setup-machine-id. 

I'd like to see the MACAddress used as the default though, this should be unique across vms, and must be for a given link anyway

**Comment by tobiaspal** [comment]
We just ran into the same problem, different instances were getting the same IP because they had the same `/etc/machine-id` file.

The [systemd machine-id documentation](https://www.freedesktop.org/software/systemd/man/machine-id.html) explains how the machine-id is derived on the first boot, and it all sounds very reasonable, and we do have different UUIDs for our KVM instances. And we don't copy the machine-id file to the instances. So how do they still end up with the same machine-id?

It turns out that we are creating the instance images with debootstrap, which installs the systemd package, and runs the systemd postinst script. And the [postinst script runs `systemd-machine-id-setup`](https://git.launchpad.net/ubuntu/+source/systemd/tree/debian/systemd.postinst#n119) in the host namespace, so the instances end up with the same machine-id as the host.

So for us the solution was to clear the `/etc/machine-id` file in the image after creating it with debootstrap.

Now how to fix the already existing instances which have the same machine-id? If I clear the `/etc/machine-id` file and run `systemd-machine-id-setup` then it gets recreated with the same, duplicate ID. [This is because there is already a valid D-Bus machine ID configured for the system](https://www.freedesktop.org/software/systemd/man/systemd-machine-id-setup.html).

There is a way to force dbus to generate a new  UUID with `dbus-uuidgen`, but according to the [`dbus-uuidgen` manual](https://dbus.freedesktop.org/doc/dbus-uuidgen.1.html), this is not a good idea.

So for already running instances, the only way seems to be to clear the `/etc/machine-id` file and restart the instance.

This does not seem to be a systemd bug. The [systemd machine-id documentation](https://www.freedesktop.org/software/systemd/man/machine-id.html) clearly states:

> For operating system images which are created once and used on multiple machines, for example for containers or in the cloud, `/etc/machine-id` should be an empty file in the generic file system image.

So in our case this was a bug in the image creation process, which did not ensure the `/etc/machine-id` file is empty in the image.

**Comment by b1tninja** [comment]
systemd-firstboot --setup-machine-id

**Comment by poettering** [comment]
I think we should just close this. THe machine ID cannot be the same on all machines. This applies here and everywhere else too. I am not convinced the ideas in https://github.com/systemd/systemd/issues/9609#issuecomment-406597061 are really the way to go, since they ultimately just hide a bigger problem.

Hope that makes sense.

**Comment by b1tninja** [comment]
The problem was that VMS or containers were being cloned without
regeneration of the machine is. The network adapter Mac address seems more
reliable here since the adapters naturally are assigned uniq addresses
